# 📊 Data Types in Python

## 📖 Introduction

Every value stored in a Python program has a specific **data type**.

A data type tells Python what kind of value is being stored and determines the operations that can be performed on that value.

For example, you can add two numbers together, but you cannot directly add a number and a string without converting one of them.

Understanding data types is essential because they form the foundation of Python programming.

## 🎯 Learning Objectives

By the end of this notebook, you will be able to:

- Understand what data types are.
- Learn why data types are important.
- Explore Python's built-in data types.
- Identify the data type of a value.
- Use the `type()` function.
- Understand mutable and immutable data types.
- Learn type checking using `isinstance()`.

## 🧠 What is a Data Type?

A **data type** defines the kind of value stored in a variable.

Python uses data types to determine:

- How the value is stored in memory.
- Which operations can be performed on it.
- How Python interprets the value.

Every value in Python belongs to a specific data type.

## 🤔 Why Do We Need Data Types?

Data types are important because they help Python:

- Store data efficiently.
- Perform valid operations.
- Prevent invalid operations.
- Improve program reliability.
- Manage memory effectively.

## 📂 Built-in Data Types in Python

Python provides several built-in data types that can be grouped into different categories.

| Category | Data Types |
|----------|------------|
| Numeric | `int`, `float`, `complex` |
| Text | `str` |
| Boolean | `bool` |
| Sequence | `list`, `tuple`, `range` |
| Set | `set`, `frozenset` |
| Mapping | `dict` |
| Binary | `bytes`, `bytearray`, `memoryview` |
| None Type | `NoneType` |

## 🌟 Did You Know?

Everything in Python is an **object**, including numbers, strings, lists, and even functions.

This is one of the reasons why Python is called an **Object-Oriented Programming Language (OOP)**.

## 🔍 The `type()` Function

The `type()` function is used to determine the data type of any object in Python.

It returns the class (type) of the given object.

### Syntax

In [1]:
type(object)

type

### Example 1: Integer

In [2]:
age = 21

print(type(age))

<class 'int'>


### Explanation

The variable `age` stores an integer value.

Therefore, `type(age)` returns `<class 'int'>`.

### Example 2: Float

In [3]:
price = 99.99

print(type(price))

<class 'float'>


### Example 3: String

In [4]:
course = "Python"

print(type(course))

<class 'str'>


### Example 4: Boolean

In [5]:
is_student = True

print(type(is_student))

<class 'bool'>


## 🧩 Numeric Data Types

Python provides three built-in numeric data types:

- **int** → Integer numbers
- **float** → Floating-point (decimal) numbers
- **complex** → Complex numbers

### Integer (`int`)
Integers are whole numbers without a decimal point.

They can be positive, negative, or zero.

In [6]:
age = 21
temperature = -5
score = 0

print(age)
print(temperature)
print(score)

21
-5
0


### Floating-Point (`float`)
Floating-point numbers (floats) contain decimal values.

They are commonly used for measurements, prices, percentages, and scientific calculations.

In [7]:
price = 99.99
height = 5.8
pi = 3.14159

print(price)
print(height)
print(pi)

99.99
5.8
3.14159


### Complex (`complex`)
Complex numbers contain two parts:

- A real part.
- An imaginary part represented by `j`.

They are mainly used in scientific computing, engineering, and mathematics.

In [8]:
number = 3 + 4j

print(number)
print(type(number))

(3+4j)
<class 'complex'>


## 🧾 Text Data Type (`str`)
A string (`str`) is used to store textual data.

Strings can be enclosed using:

- Single quotes (`' '`)
- Double quotes (`" "`)
- Triple quotes (`''' '''` or `""" """`)

In [9]:
name = "Nikita"
city = 'Jaipur'

message = """Welcome to
Python Programming"""

print(name)
print(city)
print(message)

Nikita
Jaipur
Welcome to
Python Programming


## ✅ Boolean Data Type (`bool`)
A Boolean data type has only two possible values:

- `True`
- `False`

Boolean values are widely used in decision-making and conditional statements.

In [10]:
is_logged_in = True
is_admin = False

print(is_logged_in)
print(is_admin)

True
False


## 🚫 None Data Type

The `None` value represents the absence of a value.

It is commonly used to indicate that a variable currently has no meaningful data.

In [11]:
result = None

print(result)
print(type(result))

None
<class 'NoneType'>


### 🧠 Memory Tip

Remember these five basic data types:

| Data Type | Example |
|-----------|----------|
| `int` | `10` |
| `float` | `3.14` |
| `str` | `"Python"` |
| `bool` | `True` |
| `NoneType` | `None` |

These are the data types you'll use most frequently while learning Python.

## 🔍 The `isinstance()` Function
The `isinstance()` function is used to check whether an object belongs to a specific data type.

It returns either:

- `True`
- `False`

### Syntax
isinstance(object, data_type)

In [12]:
age = 21

print(isinstance(age, int))
print(isinstance(age, float))

True
False


### Explanation

Since `age` is an integer, `isinstance(age, int)` returns `True`.

It returns `False` when checked against the `float` data type.

## 🔄 Mutable vs Immutable Data Types

Python data types are classified into two categories:

- **Mutable** → Their values can be changed after creation.
- **Immutable** → Their values cannot be changed after creation.

### Mutable Data Types

Mutable objects allow their contents to be modified without creating a new object.

Examples include:

- `list`
- `dict`
- `set`
- `bytearray`

In [13]:
fruits = ["Apple", "Banana"]

fruits.append("Mango")

print(fruits)

['Apple', 'Banana', 'Mango']


### Immutable Data Types
Immutable objects cannot be modified after they are created.

Examples include:

- `int`
- `float`
- `bool`
- `str`
- `tuple`
- `frozenset`
- `bytes`

In [14]:
name = "Python"

print(name)

Python


Although the value of an immutable object cannot be changed, you can assign a completely new value to the variable.

In [15]:
name = "Python"

name = "Java"

print(name)

Java


### 🧠 Memory Tip

- **Mutable** = Can Change ✅
- **Immutable** = Cannot Change ❌

A simple way to remember:

> **Mutable → Modify**
>
> **Immutable → Impossible to Modify**

## ⚠️ Common Mistakes

Beginners often make the following mistakes:

- ❌ Confusing variables with data types.
- ❌ Assuming `input()` returns an integer.
- ❌ Forgetting to use the `type()` function.
- ❌ Mixing strings and integers without conversion.
- ❌ Thinking that changing a variable always changes the original object.

## 💡 Best Practices

Follow these best practices while working with data types:

- ✅ Choose meaningful variable names.
- ✅ Verify data types using `type()` when needed.
- ✅ Use `isinstance()` for type checking.
- ✅ Convert data types only when necessary.
- ✅ Understand whether an object is mutable or immutable.

## 📝 Practice Questions

Answer the following questions:

1. What is a data type?
2. Why are data types important?
3. Name the three numeric data types in Python.
4. What is the difference between `int` and `float`?
5. What does the `type()` function return?
6. What is the purpose of `isinstance()`?
7. What is the difference between mutable and immutable data types?
8. What does the `None` value represent?
9. Name four immutable data types.
10. Name three mutable data types.

## 💻 Hands-on Exercises


### Exercise 1

Create variables of the following data types:

- Integer
- Float
- String
- Boolean

Print both the value and its data type using the `type()` function.

### Exercise 2

Create a variable containing your name.

Use the `isinstance()` function to check whether it is a string.

### Exercise 3

Create variables to store:

- Student Name
- Roll Number
- Percentage
- Passed Status

Display both the values and their respective data types.

## 🧩 Mini Challenge

Create a **Student Information System**.

Store the following details in variables:

- Name
- Age
- Height
- Weight
- City
- Is Student
- CGPA

Print each value along with its corresponding data type using the `type()` function.

## 🎯 Key Takeaways

After completing this notebook, you have learned:

- ✅ What data types are.
- ✅ Why data types are important.
- ✅ Python's built-in data types.
- ✅ The `type()` function.
- ✅ The `isinstance()` function.
- ✅ Mutable and immutable data types.
- ✅ Best practices for working with data types.

## 📚 Summary

Data types define the kind of values stored in a Python program.

Understanding data types helps you write correct, efficient, and reliable code.

Mastering this topic is essential because every Python program—from basic scripts to advanced Machine Learning applications—depends on proper use of data types.

## 🚀 What's Next?

In the next notebook, we will learn about **Type Casting in Python**, where you'll discover how to convert one data type into another.

➡️ **Next Notebook:** `06_Type_Casting.ipynb`